### ISSUE TITLE
Deprecation of Switch-Azuremode

### ISSUE THREAD

**Issue by trondhindenes** [Azure/azure-powershell#428]
The two "modes" in Azure are painful and confusing to work with. For example, there's two different "Get-AzureVM" cmdlets (one in each "mode"), each listing its own sets of vms (the "ServiceManagement" cmdlet lists "v1" vms and the "ResourceManager" cmdlet lists "v2" vms). This is a horrible design. Please build a PowerShell module where every module is available without having to perform confusing operations like "Switch-AzureMode": The Azure module is the only module on the planet with two modes, and for good reasons: It is a plain stupid design.

**Comment by brantb** [comment]
Running `Switch-AzureMode` now emits a warning with the message "The Switch-AzureMode cmdlet is deprecated and will be removed in a future release.", so it looks like you're getting your wish, but there doesn't seem to be any indication as to what (if anything) is replacing it.

**Comment by singhkay** [comment]
Adding @devigned. There should be more information on this in the next few days

**Comment by devigned** [comment]
Nothing has changed other than a warning, which will be followed up with some azure.com guidance (landing this week) regarding the upcoming changes and timeline for these changes.
 
We don’t expect to fully deprecate switch azure mode until after summer. Since it is such a large change, we want to start to warn users and get them thinking about what it means to no longer use switch azure mode. This change will be most impactful for users who are using both ASM and ARM cmdlets.

I'll reply back to this issue with the aforementioned guidance. Based on that guidance and timeline, we'd love to get some feedback regarding the upcoming changes and incorporate that feedback into the proposed solution.

**Comment by trondhindenes** [comment]
Thanks for the feedback,  @devigned ! Looking forward to more info on this.

**Comment by juneb** [comment]
Ogail, David, is there any guidance for partners? Our product have custom code to handle the ASM/ARM switch and I can't wait to remove it.

**Comment by markcowl** [comment]
@juneb There will be an article soon.  The practical upshot is that switching mode will no longer be necessary as you should be able to load both modules at the same time.

**Comment by devigned** [comment]
## What Is Swich-AzureMode And Why Are We Deprecating It? 
 
Beginning in version 0.8.0, the Azure PowerShell installation included both Azure Service Management (ASM) and Azure Resource Manager (ARM) modules. These modules were not designed to be used in the same session.  So to allow use of the modules within the same session, one module would be loaded and the other module would be unloaded using Switch-AzureMode. ASM was the default module loaded into the Azure PowerShell session and required the use of Switch-AzureMode to transition to ARM mode. 
 
For those who are unfamiliar with ASM and ARM, these modules represent collections of cmdlets, which target the Service Management and Resource Management REST APIs. These REST APIs provide an interface for managing Azure cloud resources such as virtual machines, networks, storage, etc. ARM and ASM can be used to create Azure resources, but the two APIs take drastically different approaches. ASM was a strictly imperative model for creating your Azure resources. ARM introduces an entirely new way of thinking about your Azure resources. Instead of creating and managing individual resources, you are able to manage entire topologies of resources together as logical units. ARM is the future of Azure management and thus the future of Azure PowerShell.  
 
Initially, ARM didn’t have all of the functionality needed to perform all of the Azure management tasks required by most users, so it was decided to default to ASM. Now, ARM is ready to be the default. Thought, rather than having a modal experience where one set of cmdlets is default, we want to load these modules side-by-side. 
 
Switch-AzureMode cmdlet was created to cover up the fact that many of the cmdlets in ARM share the same name as cmdlets in ASM (ASM/New-AzureVM and ARM/New-AzureVM), because they were used to manage the same resources (websites, vms, storage accounts, …). We really liked the names of the cmdlets in ASM and wanted to keep them for ARM, but we didn’t want to break existing Azure PowerShell scripts. Therefore, rather than changing the names of the ASM cmdlets and breaking everyone’s scripts, Switch-AzureMode was added to load one module and unload another module. 
 
The Switch-AzureMode cmdlet introduces a pattern which is weighted toward an interactive use of the cmdlets. Innately, PowerShell provides the functionality for a user to import modules and remove modules with a wide variety of options. The opinionated pattern of Swich-AzureMode has been confusing and frustrating to users and locks users into our opinionated approach to module loading.  
 
There has been significant user feedback regarding the design and we have decided to make a change for the better and deprecate Switch-AzureMode. 
Proposed Future Without Switch-AzureMode 
 
With the deprecation of Switch-AzureMode the following changes are proposed. 
- The Azure module will be renamed to “ASM”. 
- All ASM cmdlets will be prefixed with “ASM”, so Azure/New-AzureVM will become ASM/New-AsmVM or simply New-AsmVM. 
- AzureResourceManager cmdlets will keep their existing names, so New-AzureVM will be the ARM version. 
- A new convenience script will be added to provide backward compatibility to existing scripts, which only use the Azure module. The script will alias the ASM cmdlets back to their original names and remove the AzureResourceManager module from the session. 
- AzureResourceManager module will be broken into many modules by service and behavior. Examples of the module names would be the following: AzureCompute, AzureNetwork, AzureStorage, … 
- Azure PowerShell and all of its modules will be distributed via MSI and PS-Get. MSI cadence will decrease because the MSI will be a wrapper for a PS-Get script. 
 
These changes will affect a large number of users and have the potential to break many scripts. In fact, with out the use of the backward compatibility script, this change will break everyone except the users who are only using ARM cmdlets in their scripts.  

## The Timeline For These Changes 
 
With the wide-ranging impact of these changes in mind, we have decided to start talking about the transition many months before we intend on making this changes. We intend on making some of the underlying changes such as breaking up of the AzureResourceManager module into its component modules and moving to PSGet in preparation for the removal of Switch-AzureMode during July and August. The removal of Switch-AzureMode will occur in early September, which should provide ample time to prepare for these changes. 
 
![timeline-ps](https://cloud.githubusercontent.com/assets/386473/8115643/a5b522dc-1031-11e5-9cf0-8bfd311c3dd8.png)
 
## Community Feedback And Discussion 
 
Azure PowerShell is an open source project and as such, we would like to elicit feedback from the community regarding these upcoming changes. Please think of this as proposal, which is open to being shaped by feedback. This is our initial plan and is a living document. We are open to ideas from the community regarding its direction. If you feel strongly regarding this topic and would like to have your design ideas heard, please contribute to the discussion.

**Comment by trondhindenes** [comment]
Great news. Could you explain what will happen to v1/v2 vms as a result of this switch? Will "v2" vms be created by default after the switch? (that would mean that "Get-azurevm" will only list vms created via the arm api while we will have to use "get-asmvm" to list "v1" vms?). I think there's no getting around the fact that we need a unified set of cmdlets that lets us manipulate both v1 and v2 resources, or that v1 resources will be converted into v2 resources somehow. It's a mess, but that's what happens when you break an api backing a huge service such as azure....

**Comment by devigned** [comment]
@trondhindenes We will not be converting v1 to v2 resources. Transitioning from v1 to v2 resources is handled by a little differently by each Azure service. You will be able to leverage v1 cmdlets and v2 cmdlets side-by-side to manipulate v1 and v2.

**Comment by trondhindenes** [comment]
Just to get clarification: "We will not be converting v1 to v2 resources": "We" as in Microsoft, or "we" as in the team working on the PowerShell cmdlets for Azure? What I'm trying to understand here is how I should tell my customers to handle this. Say you release updated cmdlets on sept 1st that switch to arm as the default api:
august 29: vm is created with new-azurevm, which will create a V1 vm. Get-azurevm lists all vms in subscription.
september 2nd: vm is created with new-azurevm, which will create a V2 vm. Get-azurevm lists only v2 vms in subscription.

Am I correct in assuming this?

**Comment by devigned** [comment]
We as in Azure PowerShell.

The script that is calling New-AzureVM will fail as V2 expects different parameters than V1.

**Comment by jodoglevy** [comment]
## Proposed Future Without Switch-AzureMode – Azure Automation service

The Switch-AzureMode method of Azure and ARM module use did not work in Azure’s Automation service, leading to a lack of ARM adoption in the service. The above proposed changes to Azure PowerShell, along with proposed changes to Azure Automation, will make use of ASM and ARM cmdlets side by side in Azure Automation much simpler. The following changes to Azure Automation are proposed:

* Azure Automation will continue to ship the Azure module out of the box as a global module, but the proposed ASM module will also become a global module. The “split-out” ARM modules (AzureCompute, AzureStorage, etc) will also ship as global modules.
* Only the ASM and split-out ARM modules will have their global modules updated to the latest versions going forward.
* So as not to break existing Azure Automation users, existing runbooks will continue to be served the Azure module, and will not be served the ASM or ARM modules. Jobs of these runbooks will note that they are using the Azure module, that this module is now deprecated in Azure Automation, and that these runbooks should be recreated to target the ASM module instead. 
* New runbooks will not be served the Azure module, and instead will be served the ASM and split-out ARM modules.

### The Timeline For These Changes

![image](https://cloud.githubusercontent.com/assets/234589/8119393/047f584c-104a-11e5-95b6-0f68efbe581d.png)

**Comment by devigned** [comment]
@trondhindenes let me be a little more specific. The customer could also leverage a backwards compatibility script that we'll provide, which will alias the V1 commands back to their original names and remove the ARM modules. This will make the script run as it would have before, but it is only meant to be a stop gap for folks only using ASM cmdlets.

**Comment by jodoglevy** [comment]
The full proposal from Azure PS and Azure Automation teams is here: https://github.com/Azure/azure-powershell/wiki/Deprecation-of-Switch-AzureMode-in-Azure-PowerShell

**Comment by alexandair** [comment]
New Azure modules won't be offered through WebPI?
When you put new modules to PowerShellGallery, so that we can use PowerShellGet to install them, the installation path will change to c:\Program Files\WindowsPowerShell\Modules (by default) or equivalent path for the 32-bit modules. Is that correct?

**Comment by sriramvu** [comment]
@devigned , in reply to your comment "The script that is calling New-AzureVM will fail as V2 expects different parameters than V1." - 
For our service, we have planned to have same set of parameters with both V1 & V2 cmdlets (thought of not changing user experience). Can this cause any problem?

**Comment by RobinDotNet** [comment]
Let's talk about Azure Storage cmdlets. Some of these are in the ARM cmdlets, and I'm guessing that some aren't. Is that true? How will we know which cmdlet to use? Are there any in both that are the same in both?

**Comment by alexandair** [comment]
There will be a lot of cmdlets with the **same name** in ASM-based module and ARM-based modules. However, they won't have the same parameters. For example, ASM-based Get-AzureVM has one parameter set and -ServiceName parameter. ARM-based Get-AzureVM has four parameter sets and -ResourceGroupName (in two of them).

**Comment by RobinDotNet** [comment]
Yeah, I got that. I've used both for creating VMs. However, I'm specifically asking about cmdlets having to do with Azure Storage, not VMs, *especially* the ones that may not be in ARM now. Examples:
New-AzureStorageContext -- I wouldn't expect this to change
New-AzureStorageContainer -- I would expect the ASR version to require the ResourceGroup
Get-AzureStorageContainer -- might have 2 versions, with ASR version requiring ResourceGroup
Get-AzureStorageBlob  --- I wouldn't expect this one to be different
Set-AzureStorageBlobContent -- I wouldn't expect this one to be different 
Get-AzureStorageBlobContent -- I wouldn't expect this one to be different
Start-AzureStorageBlobCopy -- I wouldn't expect this one to be different 
the same ones for Files 

Robin

**Comment by alexandair** [comment]
Disregard my previous comment. I was thinking about some old discussions with Azure PowerShell team on this subject when we still talked about having the same names for both APIs and using modulequalified naming. With the current proposal, that won't be the case anymore. ASM-based module will have Get-AsmVM cmdlet and ARM-based will continue to use Get-AzureVM.

**Comment by devigned** [comment]
@sriramvu if the cmdlets exist in both ASM and ARM modules, then there should be no issues. If they will exist in a module that would be used in both, then we should talk through that so we break them up properly.

**Comment by devigned** [comment]
@RobinDotNet we should talk through breaking those cmdlets out from ASM and ARM if they have nothing to do with them. Sounds like the cmdlets you are mentioning are data plane cmdlets (non-management) and as such should be independent from the management related cmdlets (IMHO). 

Thoughts?

**Comment by alexandair** [comment]
Right now, all cmdlets that @RobinDotNet has mentioned exist in both modes--AzureServiceManagement and AzureResourceManager and are coming from **the same** DLL regardless of the mode (C:\Program Files (x86)\Microsoft SDKs\Azure\PowerShell\ServiceManagement\Azure\Storage\Microsoft.WindowsAzure.Commands.Storage.dll)
I expect them to get their own AzureStorage module and keep their name. They'll continue to work as before. For example, if I want to list all storage-related cmdlets:
```
Get-Command -Module AzureStorage
```
Are we on the same page?

**Comment by devigned** [comment]
@alexandair, yes. That's precisely what I was thinking.

**Comment by chadiel** [comment]
If we expect both cmdlets to live in the same module, would we be opening our customers to a source of frustration when they get a resource using the ASM cmdlet and try to use it with an ARM cmdlet (or vice versa) only to have the call fail? I'd imagine that a customer who has created a resource that lives side-by-side between ASM (RDFE) and ARM (e.g. SQL, Websites, etc.) to encounter this at some point (after all, from a customer standpoint they are interacting with the same resource and whether its ARM or ASM that is servicing the request is, to the customer, an implementation detail...) I worry that we may be solving one problem and introducing another...

**Comment by RobinDotNet** [comment]
@alexandair -- ALL of them are in both? But they're not identical. Get-AzureStorageAccount and New-AzureStorageAccount let you pass in a resource group in the ARM side. Does the assembly handle both? 

I'm asking because I just wrote about a thousand lines of powershell for a storage course that's going to be hosted by Opsgility (like Pluralsight, but just Azure), and it's due to be published in July, and I'm trying to figure out if I am going to have to change the code or comment about the v1/v2 issue. If the cmdlets don't change, and you guys handle where they're pulled from, that works for me.

Does that make sense?

Robin

**Comment by RobinDotNet** [comment]
One other thing. You have to deal with the v1/v2 invisibility issue. I mean where you can not see v2 storage accounts from ASM and you can't see v1 storage account from ARM. So if you load both of those, are they going to have to use Get-ASMStorageAccount and Get-AzureStorageAccount in order to see all of their storage accounts? (Considering how many millions of people are using v1 assets on Azure, this would be a p.i.t.a.)

**Comment by alexandair** [comment]
@RobinDotNet I said all cmdlets that you've listed in your comment are coming from the same DLL in both modes and will work in the future as they are working now. On the other hand, Get-AzureStorageAccount and New-AzureStorageAccount are not in the same DLL. ASM-based ones are in ```C:\Program Files (x86)\Microsoft SDKs\Azure\PowerShell\ServiceManagement\Azure\Compute\Microsoft.WindowsAzure.Commands.ServiceManagement.dll``` and ARM-based ones are in ```C:\Program Files (x86)\Microsoft SDKs\Azure\PowerShell\ResourceManager\AzureResourceManager\StorageManagement\Microsoft.Azure.Commands.Management.Storage.dll```
That means that new ASM-based ones will be renamed to **Get-AsmStorageAccount** and **New-AsmStorageAccount**. ARM-based ones will continue as Get-AzureStorageAccount and New-AzureStorageAccount.

I assume, if you want to work with v1 storage accounts, you need to use Get/New-AsmStorageAccount. Get/New-AzureStorageAccount will see v2 storage accounts.

**Comment by RobinDotNet** [comment]
You mean the old ones will be renamed to Get-Asm*, not the new ones, right? So that goes back to my original question. HOW MANY of them will be renamed? Will the ones that have the same functionality, like Get-AzureBlob, still be Get-AzureBlob, or am I going to have to change ALL of my Get-Azure* cmdlets to Get-Asm* ?

**Comment by pzerger** [comment]
@RobinDotNet - presumably the new naming convention would be applied to all the ASM cmdlets for sake of consistency. It would be confusing to have only those duplicated across modules renamed.

**Comment by pzerger** [comment]
In my view, the ARM cmdlets should also be a global module pre-loaded in Azure Automation by default (which is not currently the case). The steps to upload the ARM module (http://bit.ly/1SfgmPA) are an unnecessary distraction that should be removed for working with core Azure IaaS functionality.

**Comment by pzerger** [comment]
@devigned did you answer @alexandair question about future module distribution via WebPI? I cannot find an answer in the thread above.

**Comment by trondhindenes** [comment]
It would be great to get some statement from MS regarding how they plan to deal with migration of V1 resources to V2. I realize that's outside the scope of of the team maintaining the Azure Powershell cmdlets, but it's still a very important part of this discussion.

**Comment by tylerd** [comment]
The article mentions PowerShellGet multiple times, are you talking about PSGet? Or something different? [http://psget.net/](http://psget.net/)

**Comment by devigned** [comment]
@tylerd see: http://blogs.msdn.com/b/mvpawardprogram/archive/2014/10/06/package-management-for-powershell-modules-with-powershellget.aspx

**Comment by devigned** [comment]
@trondhindenes I'll pass on your desire to have some wider statement regarding migration from v1 -> v2

**Comment by devigned** [comment]
@alexandair and @pzerger the WebPI installer will be a facade over script running PowerShellGet. You should expect the same experience for both WebPI and PowerShellGet.

**Comment by RobinDotNet** [comment]
I never use the WebPI installer, because you have no idea what *else* is going to be installed. The last time I used it to install the PowerShell cmdlets, it updated the Azure Tools on my box, which I didn't want.

**Comment by pzerger** [comment]
@RobinDotNet last I updated my system, there is now a Standalone option in WebPI for the Azure PowerShell module I've found harmless and quick to install.

**Comment by devigned** [comment]
@RobinDotNet per your question, "HOW MANY of them will be renamed?". We don't have the total number, but I'll look into it and provide an answer. 

To provide a less concise answer, any cmdlet that is inextricably tied to ASM will be renamed. Some of the examples you listed like Get-AzureBlob are "data plane" rather than "management plane" and I would **not** consider them linked to ASM or ARM. Data plane cmdlets should not have a name change.

**Comment by devigned** [comment]
@pzerger thank you for pointing that out. We have received quite a bit of feedback similar to @RobinDotNet mentions above. We specifically added the standalone option to assuage this pain.

**Comment by RobinDotNet** [comment]
@devigned @pzerger Thanks for the info on the WebPI. I've had a lot of trouble with that thing over the years, so I try hard not to use it any more. 
@devigned That's good news. I was hoping you wouldn't rename the ones that have nothing to do with resource manager/group.

**Comment by mvanotti** [comment]
Hi!
Where can I read more about this v1/v2 resource versioning? I created VMs from the Azure Web Portal last week and I can only see them from the Get-AzureVM cmdlet from the ASM, I can't see them from the ARM one.

It is REALLY confusing to have two different ways to list resources, and I thought that it would be for legacy resources, but the vm's I'm talking about where created last week via the azure portal :/

**Comment by RobinDotNet** [comment]
I'm concerned that changing all of the v1 Powershell cmdlets is going to be a lot more confusing to the many people using them than teaching people to use Switch-AzureMode. How many people are using these -- thousands? hundreds of thousands? millions? I would bet that 99% of them have v1 assets. 

It's so clearer to me to switch to the other set of cmdlets than to figure out which cmdlet I'm going to use. Especially as noted above that not all of them will change (such as the ones from blob storage that I mentioned). So you have to figure out if the ones you use or want to use have been renamed or not. 

How did you guys decide to go in this direction? Did you happen to poll the Azure MVP distribution list and/or the Azure Insiders and ask for opinions? (If so, I didn't see it.) A lot of the MVPs do consulting and have a lot of customers, and would be able to give you some kind of prediction on how this is going to go over. 

I hate to see you do this and just get a shi*storm over it. And I'm curious what kind of input went into the decision.

**Comment by RobinDotNet** [comment]
@mvanotti You can't create v2 resources in the preview portal yet. You can only create them with the ARM cmdlets. The v1 resources and the v2 resources can not see each other, so you can't create a v1 storage account and have it used by a v2 resource (and vice versa). And not everything in v1 is in v2. I got toasted last week because it turns out the gateways are not yet available, so you can't set up site-to-site networking to the v2 resources. (And you can't put a v2 VM into a v1 virtual network.) I assume the visibility issue will always be the case, until the v1 assets are migrated to v2. You can see the v2 assets in the preview portal -- they are in separate "directories". For example, if you have v1 and v2 VMs, you will have a folder for VMs and a folder for VMs (v2).

**Comment by alexandair** [comment]
**mvanotti** The other way to get v2 resources is to use ARM templates to create compute, networking, and storage resources.

**Comment by RobinDotNet** [comment]
@mvanotti -- You're right. My last customer wanted to use PowerShell, not the templates, so I wasn't thinking about that.

**Comment by alexandair** [comment]
**RobinDotNet** The main reason for going in this direction is that Switch-AzureMode is not supported in Azure Automation. ARM is the future, and we need support for ARM-based cmdlets in Azure Automation sooner than later. No one says it will be easy. Transitional period is always PITA.

**Comment by mvanotti** [comment]
If this is already decided, then the first thing to do is add the replacement commands so now people that needs to access the legacy cmdlets can do that. I, for example, see myself using Get-AzureVM (for v1 vms) a lot, and if that function is going to be changed to Get-ASMVM, I would like to have those cmdlets available sooner rather than later.

**Comment by mvanotti** [comment]
@RobinDotNet, @alexandair thanks for your answer, but I'm looking at docs that describes the new things about v2 stuff, is there such thing documented?

**Comment by RobinDotNet** [comment]
@mvanotti Hahaha! That's funny. Hahaha! I mean, no. There's nothing about the visibility or about what features aren't available. My customer had to revert to v1. After discovering the network gateway lack, he was unwilling to continue, afraid we would hit something else that was missing. 

Here are some helpful links I found.

I had considerable trouble creating my customer's infrastructure in PS (2 vm's in a load balancer and availability set). This shows how to do that, but doesn't explain why you're suddenly configuring nics and load balancers, which you didn't have to/couldn't do in v1. 

https://azure.microsoft.com/en-us/documentation/articles/virtual-machines-create-windows-powershell-resource-manager/

Also helpful was Neil Mackenzie’s blog article:

http://blogs.msdn.com/b/cloud_solution_architect/archive/2015/05/05/creating-azure-vms-with-arm-powershell-cmdlets.aspx 

and most of all, his gist: https://gist.github.com/nmackenzie/042bdffdfa6098f5063a 

This is about the templates:
https://azure.microsoft.com/en-us/documentation/articles/virtual-machines-deploy-rmtemplates-powershell/

And this is from Corey Sanders talking about the differences.
https://azure.microsoft.com/en-us/documentation/articles/virtual-machines-azurerm-versus-azuresm/

And this is from Ignite. Start at 7 minutes to learn about the different object models.
http://channel9.msdn.com/events/Ignite/2015/BRK3505 

Good luck to us all.
Robin

**Comment by RobinDotNet** [comment]
@alexandair Ohhhh, I see. It's going to be a huge impactful change that's probably going to make some people angry. I wish Microsoft luck with it, and hope it works out okay.

**Comment by jodoglevy** [comment]
@pzerger this solution is exactly to address the problem you mention, of ARM cmdlets not being pre-loaded in Azure Automation. See my post above for details.

**Comment by pzerger** [comment]
Joe, caught that on re-read top to bottom. Seems like the right plan to me.

Regards,

Pete Zerger, MVP
Founder, SystemCenterCentral.com
Azure IaaS Book (coming soon) - http://insidethemicrosoftcloud.com
On Mon, Jun 15, 2015 at 7:31 PM, Azure/azure-powershell
<reply@reply.github.com>
wrote:
@pzerger [https://github.com/pzerger] this solution is exactly to address the problem you mention, of ARM cmdlets not
being pre-loaded in Azure Automation. See my post above for details.

—
Reply to this email directly or view it on GitHub
[https://github.com/Azure/azure-powershell/issues/428#issuecomment-112244192] .[https://github.com/notifications/beacon/AJ0QC-5UIldF5iWH926kXQyedFkLvCRxks5oT2XhgaJpZM4Eqr2X.gif]

**Comment by devigned** [comment]
@mvanotti 
> If this is already decided, then the first thing to do is add the replacement commands so now people that needs to access the legacy cmdlets can do that. I, for example, see myself using Get-AzureVM (for v1 vms) a lot, and if that function is going to be changed to Get-ASMVM, I would like to have those cmdlets available sooner rather than later.

I wouldn't worry about making replacement commands. We intend on providing backwards compatibility for users using only ASM. I wasn't clear on exactly how b/c we are still talking through some of the options (script to rename ASM cmdlets to their new names, aliasing ASM cmdlets script w/ Remove-Module for ARM which keeps their old names, or other options).

With the aliasing ASM cmdlets script w/ Remove-Module option, we'll run into issues with scripts being used in PowerShell workflow and Azure Automation  (see: https://technet.microsoft.com/en-us/magazine/dn151046.aspx).

**Comment by devigned** [comment]
Please keep in mind that this is a proposal. We are open to ideas. If anyone feels strongly about this one way or another, let's try to come to the best solution for the Azure PowerShell community.

**Comment by trondhindenes** [comment]
I'm trying to think of all this in terms of scenarios. I have lots of customers with various types of Powershell-based automation doing stuff in Azure. For these I will simply advise them not to update the cmdlets until everything is tested and ready, so that's not a huge problem.

My worry is that customers who want to move forward/migrate to v2 resouces are met by a ton of problems because of not only breaking changes but the requirement to use different cmdlets to access their existing (v1) resources and their newly created (v2) resources. I'm worried that this will lead to customers being reluctant to use all the new functionality that v2 represents, because they basically break their automation once the first v2 resource is created. 

Again, I must ask Microsoft to provide guidance around how to convert/transfer/upgrade v1 resources to v2. The discussion around v1/v2 cmdlets isn't going to go anywhere until we know more about what will happen to the resources themselves.

**Comment by mvanotti** [comment]
Hi!

Thank you all for your answers.
My understanding is that current customers that are using v1 might keep the same resources in v1, and maybe add new resources from v2. So, it would be nice to have a way of interacting with both versions easily (not having to load-unload a module, or even switching modules), that's why I thought having the ASM cmdlets renamed was a good solution for users of both versions.

**Comment by singhkay** [comment]
@trondhindenes The team is currently working through planning on how to provide an automatic migration solution. Here's a survey that was sent out last week to gather feedback and requirements to better understand customer scenarios - http://aka.ms/armmigration

**Comment by jodoglevy** [comment]
@trondhindenes Different Azure services have different backwards compatibility "amounts" in their Azure Resource Manager (ARM) APIs. With Azure Automation's ARM API, for example, you can fully manage resources created through Azure Service Management (ASM) using ARM. As far as I know, Azure Compute's ARM functionality at this time cannot interact with ASM compute resources, but @singhkay can speak more to that. Other Azure services I cannot speak for.

But the backwards compatibility of any ARM functionality for managing ASM resources, in my opinion, should be addressed with the Azure teams that are not providing the backwards compatibility needed. Azure's PowerShell team is responsible for exposing all ASM and ARM functionality via PowerShell cmdlets -- the team does not have the ability to change which Azure services' APIs support backwards compatibility, that is the responsibility of each Azure service team.

Azure is in a transition period from ASM to ARM, and so regardless of ARM's backwards compatibility, customers will still want to use a combination of ASM and ARM functionality in the near future, even if its just because they're currently comfortable with ASM. As @mvanotti said above, exposing all the ASM and ARM functionality as cmdlets, in a way where switching, loading/unloading, and namespacing are not needed, should allow for the easiest use of whatever cmdlets customers need to get their tasks automated.

**Comment by MattAllison** [comment]
I can't possibly begin to explain how frustrating this is.

It seems that the decision is being made to throw bad engineering after bad engineering; introducing the Switch-AzureMode to allow the re-use of commands with different parameters and methods of use because "we liked the name".

At least for me, September is an insane timeframe to re-work scripts that took 6 months of professional services time to develop even if the "work" is to go back and add module version checks.

It seems to me that the right approach is to learn from the Switch-AzureMode debacle rather than build on it.  Admit that the "nice names" isn't a sufficient reason to disrupt your customers and use new names like verb-AzureRMnoun for the new activities.

**Comment by devigned** [comment]
@MattAllison, I love the passion. Using verb-AzureRM for the ARM cmdlets is certainly an option and is something we have discussed.

This approach does introduce a couple things that could be a positive or negative base on point of view. 

- With all Azure cmdlets having the verb-Azure* structure, it might be a little crowded when searching for Azure cmdlets. With the verb-Asm and verb-Azure naming, you can separate ASM and ARM a little easier.
- I think users will have a tendency to view verb-Azure cmdlets as the "default" and verb-AzureRM as the alternative. The future is ARM, so it we'd like to encourage users to move to ARM even through small aesthetic cues. 

Do you think either of these points provide value?

What say you, others following this issue?

**Comment by MattAllison** [comment]
"Passion" is a very generous term; thanks for that.

It's true that with all of the cmdlets starting with verb-Azure searching could be crowded, but it would also allow you to see items that have direct new comparative commands fairly close to each other (get-command *-Azure*Storage)   Yes, that has to be learned, but so does all of this and I think it's less of an impact.

Users may have a tendency to view the legacy commands as the default but that's pretty easily overcome with training / communication as well.  I think it would work itself out very quickly once people couldn't create VMs on their networks because the underlying bits are fundamentally different.  I think you could also add some checking to the old/new commands that would say something like "The resource you targeted exists but is the wrong version to be used with this command - use <alternative> instead".

I'm going to go out on a limb and guess that when the first version of Azure cmdlets were developed, Azure Resource Manager wasn't really part of the consideration.  Likewise, I'm going to assume that Azure Resource Manager may not reflect a future generation of Azure and that similar changes may occur again.  If I'm even close to right on the latter part this current approach doesn't set a framework that avoids this type of problem recurring.

**Comment by NaveenGoli** [comment]
How about an option where -v1 will run ASM cmdlets and -v2 will run the ARM cmdlets?

**Comment by MattAllison** [comment]
I haven't looked @ all of the cmdlets before writing this but I'd think that a parameter set would be a better approach.  If -serviceName is being used then use ASM mode, if -resourceGroup is being used then use ARM mode.

**Comment by alexandair** [comment]
**@MattAllison** That might look like an elegant solution, but keep in mind that ASM- and ARM-based cmdlets are based on different APIs.
For example, ASM-based Get-AzureVM has **2** parameter sets at the moment. **Only one** of them has the -ServiceName parameter.
ARM-based Get-AzureVM has **4** parameter sets at the moment. **Only 2** of them have the -ResourceGroup parameter. You cannot use the -ServiceName and -ResourceGroup parameters to differentiate ASM- and ARM-based cmdlets.

**Comment by mvanotti** [comment]
I am working on a PowerShell Module, that depends on the Azure Module.
If I put it as a dependency on my PSD file, I get an error while running Switch-AzureMode in my cmdlets

"WARNING: The Switch-AzureMode cmdlet is deprecated and will be removed in a future release.
Switch-AzureMode : Unable to remove the module 'Azure' because it is required by 'MyAzureModule'. Add the Force parameter to 
your command to remove the module."

**Comment by MattAllison** [comment]
Regarding the parameter sets, it's true that the ASM version has two options.  It still seems easier to me to resolve those types of items by either requiring a new parameter (-ASM -ARM) or by determining which mode to use based on what resources are available (ASM mode if there aren't ARM resources, asking for a specific parameter if there are both).   Personally I think it makes more sense to simply require the -All parameter for the ARM resource and leave ASM behavior unchanged.

Regardless, my original point stands; It seems to make more sense to leave the current ASM intact rather than forcing all existing customers (and me specifically - this is largely self-interest) hundreds of hours of work to retool scripts.

**Comment by jodoglevy** [comment]
Matt, can you explain why this would require hundreds of hours of rework? The only thing changing about the ASM cmdlets is their names (Get-Azure* -> Get-Asm*). If we provided a script you could run against your existing scripts, that would change all Get-Azure* references to Get-Asm* in them, would this not suffice?

**Comment by jodoglevy** [comment]
(And all non-get cmdlets as well, of course)

**Comment by MattAllison** [comment]
Part of it is changing the master copies of our fully developed scripts and then coordinating their deployment along with the update of the Azure PowerShell module across workstations and the servers where some of the scripts run unattended or as part of other workflow processes.

Not all of our scripts are "complete" scripts - we also have a library of code snippets for reference and tweaking of one-off activities.  There is also documentation and internal training material we have to update.

Also, some of our scripts were developed for us by a 3rd party (Microsoft ironically) and there would be out-of-pocket real money costs to re-engage them if our internal staff isn't comfortable changing the code.

**Comment by trondhindenes** [comment]
@jodoglevy I think the worry is that scripts need to be (significantly) rewritten in order to utilize v2 resources (or them v1 resources are retired, which we still haven't gotten any information on ).

From my point of view, the v2 resources have a lot of features I'd like my customers to use, but once they introduce the first v2 vm in the mix, all housekeeping scripts are basically useless as they are today, and need a significant rewrite in order to "see" both v1 and v2 resources.

I keep repeating myself, but this discussion is useless without some info from Microsoft regarding:
1. How will we transistion from v1 to v2 resources
2. How long will v1 resources be supported (and will they get auto-upgraded?)

**Comment by jeetgohil814** [comment]
If I'm creating a new Azure PS script from scratch, Is there anyway not to use the Switch-AzureMode cmdlet right now so I'll be fine once the switch occurs? Or will I have to do it this way, and then either fix it in September, or use backwards compatibility then?

**Comment by MattAllison** [comment]
To add to what @trondhindenes has mentioned.  The work to modify existing scripts to allow ASM work to continue is additive work; we still have to create our own new versions that work with ARM.  It's double work.

**Comment by jodoglevy** [comment]
Thanks @MattAllison. @trondhindenes as I said above Azure PowerShell doesn't work on the compatibility of ARM and ASM (RDFE) APIs, that is up to service teams. Azure's PowerShell team just creates a wrapper around the ARM and ASM APIs, whatever they may be, and whatever compatibly between v1 and v2 they have. You'll have to ask those questions around how to transition v1 resources to v2, and how long v1 resources will be supported, to the teams that seem not to provide a path forward from v1 to v2. That is a different discussion.

This discussion is about how to use the ASM and ARM PowerShell cmdlets side by side. ASM and ARM cmdlets take different parameters and in many cases provide different functionality (one example is v1 vs v2 resource use for some services, as you say), so discussing how to use these cmdlets together, regardless of whether the APIs behind the cmdlets support v1 and/or v2 resources, is needed.

**Comment by joshuapetryk** [comment]
Any solution to @jeetgohil814 's question? We are going to be writing some ARM templates but I hate to use Switch-AzureMode in new scripts. What are we supposed to use as an alternative in the meantime?

**Comment by devigned** [comment]
@MattAllison I've been thinking about your feedback (and others) for a bit, and it resonates with me.
>[MattAllison] It seems to me that the right approach is to learn from the Switch-AzureMode debacle rather than build on it. Admit that the "nice names" isn't a sufficient reason to disrupt your customers and use new names like verb-AzureRMnoun for the new activities.

>[MattAllison] Regardless, my original point stands; It seems to make more sense to leave the current ASM intact rather than forcing all existing customers (and me specifically - this is largely self-interest) hundreds of hours of work to retool scripts.

Let's say that we choose to rename the ARM cmdlets and leave the ASM cmdlets alone. What would be the best noun prefix for those cmdlets? AzureRM, AzureV2, something else? Anyone have any ideas?

Upon reflecting on these changes, it seems like changing the ARM cmdlets will satisfy our existing commitment to users preserving ASM at the cost of losing the Azure noun prefix. This seems like a reasonable price to pay in the present, which will pay dividends if Azure ever introduces an ARM replacement in the distant future.

I'm interested in hearing how the community feels about this change to the proposal.

**Comment by trondhindenes** [comment]
I agree. Moving the arm cmdlets to the "*AzureRM*"-prefix will cause the least amount of disruption of current scripts. I would move quickly tho, as some customers have no doubt already started using arm cmdlets.

**Comment by pzerger** [comment]
@jodoglevy renaming the ARM cmdlets was something I suggested on User Voice back in May (http://feedback.azure.com/forums/246290-azure-automation/suggestions/6634825-enable-azureresourcemanager-module-in-azure-automa) as well as our private thread with the PowerShell MVPs. Looping in @alexandair for a reminder if there was any reason this was undesirable.

**Comment by pcgeek86** [comment]
@devigned @jodoglevy @markcowl @singhkay @ogail PowerShell has solved the conflicting command name issue since the beginning. The proper syntax to call a PowerShell command is:

`<ModuleName>\<CommandName>`

```
### Create a VM
AzureResourceManager\New-AzureVM
Azure\New-AzureVM

### Create a Storage Account
AzureResourceManager\New-AzureStorageAccount
Azure\New-AzureStorageAccount
```

If you want to reduce the confusion and impact to customers, then separate the modules, and educate customers about how to properly call into each module. This is not a technical problem. This is an educational problem, which I have discussed with most of you at length.

Cheers,
Trevor Sullivan
Microsoft MVP: PowerShell

**Comment by MattAllison** [comment]
I appreciate Trevor's perspective but am unsure how a non-pathed method call would be intercepted.  Would "Azure\" win or would "AzureResourceManager\" win if there was a cmdlet with the same name that existed in both.

Because we will have a hybrid environment for an unknown period of time, it would seem that this does not solve the need to go back and modify code.  This type of change wouldn't be acceptable in compiled code, I don't know why we should have to accept it for scripting.

Amazon has done a reasonable job at separating their commands (http://docs.aws.amazon.com/powershell/latest/userguide/pstools-discovery-aliases.html) which may be a pattern to follow if the "-AzureRM" seems to verbose.  In this case, you could implement -ARM and -ASM as parallel new code and freeze "-Azure" functions and parameters as they are now.

This may also allow you to have some shorter names for other commands; Start-AzureSiteRecoveryreallylongcommandname can become Start-ASRsomewhatshortername and likewise the *-AzureStoreSimple* commands can become *-ASS* (not a deliberate editorial on the product, but amusing).

**Comment by trondhindenes** [comment]
I'm not sure I agree with the notion of identical cmdlets in different modules, Trevor. Imagine all the confusion for people trying to get help in forums etc. I'm not opposed to shortening down the naming tho, some of them terribly long.

**Comment by pzerger** [comment]
I can tell you ARM loses today as it's not the default mode.

Regards,

Pete Zerger, MVP
Founder, SystemCenterCentral.com
Azure IaaS Book (coming soon) - http://insidethemicrosoftcloud.com
On Mon, Jul 13, 2015 at 11:01 AM, MattAllison <notifications@github.com> wrote:
I appreciate Trevor's perspective but am unsure how a non-pathed method call
would be intercepted. Would "Azure\" win or would "AzureResourceManager\" win if
there was a cmdlet with the same name that existed in both.

Because we will have a hybrid environment for an unknown period of time, it
would seem that this does not solve the need to go back and modify code. This
type of change wouldn't be acceptable in compiled code, I don't know why we
should have to accept it for scripting.

Amazon has done a reasonable job at separating their commands ( 
http://docs.aws.amazon.com/powershell/latest/userguide/pstools-discovery-aliases.html
[http://docs.aws.amazon.com/powershell/latest/userguide/pstools-discovery-aliases.html] ) which may be a pattern to follow if the "-AzureRM" seems to verbose. In this
case, you could implement -ARM and -ASM as parallel new code and freeze "-Azure"
functions and parameters as they are now.

This may also allow you to have some shorter names for other commands;
Start-AzureSiteRecoveryreallylongcommandname can become
Start-ASRsomewhatshortername and likewise the -AzureStoreSimple commands can become -ASS (not a deliberate editorial on the product, but amusing).

—
Reply to this email directly or view it on GitHub
[https://github.com/Azure/azure-powershell/issues/428#issuecomment-120979524] .[https://github.com/notifications/beacon/AJ0QC5mpR8jJfJd_93OokjdtHVMZ11R9ks5oc9hOgaJpZM4Eqr2X.gif]

**Comment by MattAllison** [comment]
If we went with the "rename everything" option, we would get away from identical commands in different modules and setup a path for future similar problems when a hypothetical ARM upgrade/successor comes out with more breaking changes.

**Comment by RobinDotNet** [comment]
I kind of like the “rename everything” option. It forces people to specify which one they are using, and removes the chance that they will try to use the wrong one and not realize it. Plus, as noted below, it sets up a future path for similar changes.

Robin

**Comment by pzerger** [comment]
Agreed. I don't think replacing the rename approach with A massive educational
undertaking with regards to pathing is a tenable approach.

Regards,

Pete Zerger, MVP
Founder, SystemCenterCentral.com
Azure IaaS Book (coming soon) - http://insidethemicrosoftcloud.com
On Mon, Jul 13, 2015 at 11:18 AM, RobinDotNet <notifications@github.com> wrote:
I kind of like the “rename everything” option. It forces people to specify which
one they are using, and removes the chance that they will try to use the wrong
one and not realize it. Plus, as noted below, it sets up a future path for
similar changes.

Robin

**Comment by devigned** [comment]
@pzerger @RobinDotNet I'm trimming your email responses to keep the thread readable.

**Comment by devigned** [comment]
@pcgeek86 the thing about having the same name across modules that worries me is the way PowerShell loads modules in the path. See the following example from @jodoglevy. In the case where both modules which contain a cmdlet are in the path the following is likely to occur.

> I started playing around with this. First I wanted to test what Aleksandar said about how the “default” assumed module would switch. I have validated that is correct:

> I created a fake Azure (“AzureServiceManagement”) and ARM module, both with a cmdlet “Test-Cmdlet” in it, that outputted the module the cmdlet was in:
 
> You can see these are indeed in the PSPath:
```PowerShell
PS C:\windows\system32> Get-Module -ListAvailable
Directory: C:\Users\someuser\Documents\WindowsPowerShell\Modules
ModuleType Version    Name                                ExportedCommands                                                                 
---------- -------    ----                                ----------------
Script     0.0        AzureResourceManager                Test-Cmdlet
Script     0.0        AzureServiceManagement              Test-Cmdlet                                  
```

> Without running import-module, ARM module is the default, since it comes first alphabetically for this folder in PSPath:
```PowerShell
PS C:\windows\system32> Test-Cmdlet
I am in the AzureResourceManager module
```
>Namespacing the ASM module correctly uses that cmdlets:
```PowerShell
PS C:\windows\system32> AzureServiceManagement\Test-Cmdlet
I am in the AzureServiceManagement module
```
> But now the default for the rest of the session is ASM, since now this module has been loaded last, and last loaded default trumps first in PSPath when not loaded default. No way to switch if back without explicitly doing import-module:
```PowerShell
PS C:\windows\system32> Test-Cmdlet
I am in the AzureServiceManagement module
PS C:\windows\system32> AzureResourceManager\Test-Cmdlet
I am in the AzureResourceManager module
PS C:\windows\system32> Test-Cmdlet
I am in the AzureServiceManagement module
```
> This is clearly not what we want.

This illustration of PowerShell module loading is far from intuitive and something that I don't think most people will understand with out groking PowerShell module loading as a whole. This seems like it will put a high bar for entrance into the community.

**Comment by pcgeek86** [comment]
The .NET platform allows "duplicate" type names, as long as they exist in different namespaces. Think of a PowerShell module like a namespace. Look at the `System.Windows` and `System.Windows.Forms` namespaces. There are **tons** of duplicated class names, .NET enumerations, etc.

https://msdn.microsoft.com/en-us/library/system.windows(v=vs.110).aspx
https://msdn.microsoft.com/en-us/library/system.windows.forms(v=vs.110).aspx

```
System.Windows.Forms.Application
System.Windows.Application
```

^^^ This isn't confusing for .NET developers, so why would the following be confusing for PowerShell developers?

```
Azure\Get-AzureVM
AzureResourceManager\Get-AzureVM
```

The problem we're discussing here is quite straightforward, and PowerShell has a built-in mechanism to handle it. Why are you trying to deviate from PowerShell norms? That is what confuses me.

Cheers,
Trevor Sullivan
Microsoft MVP: PowerShell

**Comment by pcgeek86** [comment]
I get the impression that people are trying to protect other people from themselves, rather than educating them. That is a scary approach, but it seems that is what the masses want.

Cheers,
Trevor Sullivan
Microsoft MVP: PowerShell

**Comment by MattAllison** [comment]
I think this is where there is perhaps a MASSIVE disconnect with a lot of what Microsoft is doing in Azure.

In my experience, the majority of PowerShell users are NOT developers and.  They don't have the background, experience or skills to deal with hierarchal namespaces of methods and classes, don't understand strongly typed variables, JSON, XML, etc..  And many or most of them never truly will.

Fact.  Sad but true.

You need to distinguish between PowerShell developers (those of a higher level of skill that may do things like write PowerShell modules and wrappers) and PowerShell users who write simple batch-file equivalents that probably don't use parameters effectively and nobody will ever remember how they work a month after they're created but know that they're important and that they do this because they have some other job to do that probably has nothing to do with actually writing PowerShell; it's a tool, not a lifestyle and it's probably viewed as being in their way.

In short, Azure PowerShell has to be reachable by PowerShell users and development and growth has to be as well.  I'd be happy to take this point offline (Azure Advisors works) and give you some more specific examples than I can in a public setting if that would help.

I am DEFINATELY not saying not to train people.  But don't make it harder to train them by obsoleting their skills every 9 months and, yes protect them (and us) from themselves while they learn.

**Comment by pcgeek86** [comment]
I agree about obsoleting skills every 9 months. Rather than us arguing about a long-standing (10 year) PowerShell paradigm, let's join forces to focus improvement of major Azure PowerShell design inconsistencies, such as:

* The IaaSv2 (ARM) Virtual Machine image commands (5 instead of 1)
* The broken ARM `Get-AzureLocation` command
* The ARM `Set-AzureSubnet` command requires you specify the `-AddressPrefix` parameter on an existing subnet, even if you're not changing it
* ... and I'm sure many more

These are **much** more frustrating to me than having to prefix my PowerShell commands with the appropriate module name. It's nearly impossible to keep up with this stuff, as evidenced by Pete Zerger's (and my own) recent comments on Azure Advisors.

https://www.yammer.com/azureadvisors/#/Threads/show?threadId=533631716
https://www.yammer.com/azureadvisors/#/Threads/show?threadId=560232221
https://www.yammer.com/azureadvisors/#/Threads/show?threadId=535991876
https://www.yammer.com/azureadvisors/#/Threads/show?threadId=553095471

Let's join forces rather than arguing about a 10 year-old precedent. Can we agree on that? :) 

Cheers,
Trevor Sullivan
Microsoft MVP: PowerShell

**Comment by MattAllison** [comment]
I agree we should join forces and push for the correct solution.  I'm not arguing the paradigm -- I'm hopefully pointing out that the implementation, while working within rather than around the architecture of PowerShell, must readily support IT staff without developer skillsets.  The paradigm supports pathed namespaces; using that is difficult and shouldn't be part of the implementation.

I wholeheartedly agree with you on the ARM commands being significantly more difficult to deal with.   While I don't have my list in front of me, this includes the needless complexity of steps to specify the image and handling network interfaces, repetitive prompting for ResourceGroup and/or other parameters.

Sadly, I hadn't really been able to bring these things up earlier.  I probably should have done that in parallel rather than solely focusing on the cutover/timeline problem.

**Comment by devigned** [comment]
Yes, each of the issues @pcgeek86 enumerated should be addressed, and it's great to bring them to light for the community to rally behind. Each should be their own issue and have a discussion for each of them.

This thread has had multiple orthogonal discussions spawn. If we can, please try to keep the focus for this thread on the topic at hand, the proposed structure of Azure PowerShell post Switch-AzureMode.

**Comment by pcgeek86** [comment]
@devigned Good deal, well I think we know where my point stands. I would highly recommend proposing the designs to the PowerShell core team and PowerShell MVPs prior to moving forward with any particular direction.

**Comment by jachin84** [comment]
The comment from @MattAllison resonates with me as well.
>"Admit that the "nice names" isn't a sufficient reason to disrupt your customers and use new names like verb-AzureRMnoun for the new activities."

My thoughts are:
* There is huge investment in the current ASM Cmdlets, don't break them!
* A separate name for ARM Cmdlets is simple solution and simple is always good.
* I like *verb*-ARM*Noun*

**Comment by trondhindenes** [comment]
@jachin84 I competely agree.

**Comment by marcusrobinson** [comment]
Can a -ARM or -ASM switch not be added to the current commands? If a switch isn't provided then defaults to ASM so existing scripts do not break. If a switch isn't provided give a warning to explain one should be provided?

Always the option to change the default at later point in time. Also if another change is made in the future just add another option, I.e. -AxM, otherwise this could be in this situation again.

**Comment by doctordns** [comment]
Let's face it, the Two-headed API monster that is Azure is not somethign anyone would have designed from the start. From an IT Pro's point of view,  you can do the same things in each API, just in an utterly incompatible way. It's madness - and I think Matt Allison hit the nail on the head with his comment above about IT pros not being developers!  But moving on - I think we are between a bit rock and a huge hard place. The existing schizophrenic nature of the Azure module is just plain wrong from a PowerShell point of view and from the understanding of the IT Pros I talk about. It needs to change - although ideally if it had all been designed right from the start we would not be in this situation. But it wasn't and we are. 

In an ideal world we would have just one API and just one set of PowerShell cmdlets. Is there a plan to really move toward this approach? Or is the real plan to keep this level of confusion alive. From this thread I am confused more than ever on where Azure PowerShell is going?

So moving forward - what is the plan of record? Is this Great Renaming going to take place? I just want to be able to create modules (like my AzureHelper module here in GitHub) that does not have to be re-written ever so often just because we have yet another new API, or another cmdlet re-design. I do not think it too much to ask for someone in charge to make a long term decision and give the IT Pro community some sensible stability.

**Comment by trondhindenes** [comment]
+1 and amen to that, @doctordns

**Comment by pzerger** [comment]
+1 definitely time to move on a plan now before IaaS v2 sees a wide
adoption and we have two large groups (instead of one) worried about
potential breakage of their deployment scripts.

Regards,

Pete Zerger, MVP (System Center Cloud and Datacenter Management)
Founder, System Center Central: http://www.systemcentercentral.com
Azure IaaS Book (coming soon) http://InsideTheMicrosoftCloud.com

**Comment by robplank** [comment]
Last I heard they are still on schedule to meet the timeline Joe Levy posted in his comment on this thread on June 11th.

**Comment by MattAllison** [comment]
Which seems insane to me considering that (AFAIK) Express Route still doesn't work with ARM VNets and there is no migration story to move from ASM to ARM that doesn't involve rm -rf *.

**Comment by trondhindenes** [comment]
According to this, they were supposed to provide some info last or this week. Which means the have about 24 hours do do it: https://social.msdn.microsoft.com/Forums/azure/en-US/2e492d2d-41eb-47ef-8818-caa4581a652c/transition-existing-vmsresources-into-arm?forum=WAVirtualMachinesforWindows

**Comment by jodoglevy** [comment]
@devigned and I have closed on a plan that we will officially post out soon, same timelines as before. Short version - ARM cmdlets will be renamed, ASM cmdlets will keep the same names. This will allow usage of ARM and ASM cmdlets side by side without breaking scripts that use ASM cmdlets. Scripts using ARM cmdlets will break, but only when you personally decide to move to the latest version of Azure PowerShell.

Again, this discussion is solely about using ARM and ASM PowerShell cmdlets side by side. ARM's ability, whether through UI, PowerShell, REST API, etc, to manage resources created by RDFE (ASM), is a separate issue. If an Azure service is not providing adequate management of ASM-created resources through ARM, then that issue needs to be addressed with the team responsible for that Azure service.

**Comment by MattAllison** [comment]
@jodoglevy Thank you very much for the update and, while I'm sure that's a bit of a pain from the development side, it sounds like you've made the right decision for your customers.

**Comment by doctordns** [comment]
@jodoglevy thanks for the update. Renaming the ARM vs ASM cmdlets makes a great deal of sense. Its a MUCH better idea, imho, than the previous plan. This saves me, personally, a great deal of work. 

I think we all hear you about the other issues (beyond ARM vs ASM cmdlet, their names, etc). My comments above still lagely stand - this is a large source of confusion, especially for IT Pros. We need some executive clarity over the future or ARM vs ASM. Issues sucyh as @MattAllison mentions just shouldn't have happened. Given that they HAVE happened, someone at MS needs to step up to the plate with some sort of sensible strategy. It feels like, at the moment, strategy is being made up as Azure goes along - we need a decent long term plan and miles stones against that plan.

**Comment by doctordns** [comment]
A further question - can we take @jodoglevy's post above as definitive - in other words is this Microsoft's final decision?

**Comment by devigned** [comment]
@doctordns this is the official decision after weighing all of the feedback from the initial proposal. 

I'll be updating the proposal to reflect the changes very soon.

**Comment by doctordns** [comment]
@devigned Thanks. As I said above, I think this is the best of some uncomfotable choices. I look forward to it comign about swiftly and smoothly. While saying this, there is a major communication exercise to undertake - This is going to affect at least some existing scripts - and the IT Pros behind those need forewarning!

**Comment by devigned** [comment]
[See wiki article for previous edits](https://github.com/Azure/azure-powershell/wiki/Deprecation-of-Switch-AzureMode-in-Azure-PowerShell)

## Executive Summary
### What is Changing:
Azure PowerShell team is removing Switch-AzureMode, which will be a breaking change to create a solid foundation for the future of Azure PowerShell. The changes are as follows and are targeted for release on 9/25/2015.
- Remove Switch-AzureMode remove modal behavior from Azure PowerShell
- Decompose AzureResourceManager module into component modules by service and functionality (management vs data plane by service)
- Rename cmdlets in AzureResourceManager module from [Verb]-Azure[Noun] to [Verb]-AzureRM[Noun]
- Distribute Azure and AzureResourceManager modules via PowerShell Gallery
- Adding automated documentation submission to content team for publication (MSDN) upon release for all modules

### Why it is Changing:
Azure PowerShell has received strong criticism for its modal / stateful behavior, which is rooted in the use of Switch-AzureMode to switch between Azure Service Management and Azure Resource Management modes. These changes will cause a breaking change, but will provide a strong foundation for the advancement and adoption of Azure PowerShell moving forward.

### Result of Changes:
- Azure PowerShell will be stateless and enable simple use of ARM and ASM.
- Azure service teams can release module quicker on their own cadence.
- Scripts can run unchanged in Azure Automation.
- Documentation will be up-to-date with each release.
- Breaking change for all users currently targeting Azure Resource Manager.
- Content on Azure.com, MSDN and others needs updated prior to 9/25/2015.

## What Is Switch-AzureMode And Why Are We Deprecating It?

Beginning in version 0.8.0, the Azure PowerShell installation included both Azure Service Management (ASM) and Azure Resource Manager (ARM) modules. These modules were not designed to be used in the same session.  To allow use of the modules within the same session, one module would be loaded and the other module would be unloaded using Switch-AzureMode. ASM was the default module loaded into the Azure PowerShell session and required the use of Switch-AzureMode to transition to ARM mode.

For those who are unfamiliar with ASM and ARM, these modules represent collections of cmdlets, which target the Service Management and Resource Management REST APIs. These REST APIs provide an interface for managing Azure cloud resources such as virtual machines, networks, storage, etc. ARM and ASM can be used to create Azure resources, but the two APIs take drastically different approaches. ASM was a strictly imperative model for creating your Azure resources. ARM introduces an entirely new way of thinking about your Azure resources. Instead of creating and managing individual resources, you are able to manage entire topologies of resources together as logical units. ARM is the future of Azure management and thus the future of Azure PowerShell. 

Initially, ARM did not have all of the functionality needed to perform all of the Azure management tasks required by most users, so it was decided to default to ASM. Now, ARM is ready to be the default. Though, rather than having a modal experience where one set of cmdlets is default, we want to load these modules side-by-side.

Switch-AzureMode cmdlet was created to hide the fact that many of the cmdlets in ARM share the same name as cmdlets in ASM (Azure/New-AzureVM and AzureResourceManager/New-AzureVM), because they were used to manage the same resources (WebApps, VMs, Storage Accounts …). We preferred the names of the cmdlets in ASM and wanted to keep them for ARM, but we did not want to break existing Azure PowerShell scripts. Therefore, rather than loading both modules and relying on PowerShell load order, one module was loaded at a time creating a modal / stateful experience for Azure PowerShell.

Switch-AzureMode cmdlet introduces a pattern, which is biased toward an interactive use of the cmdlets. When Switch-AzureMode is called, the module path is changed to include the selected cmdlets and previous mode’s cmdlets are removed from the path. This activity is stateful and causes other scripts running outside of the current session to be affected by this change. This is unexpected functionality and is quite troublesome when running multiple Azure PowerShell sessions.

PowerShell provides the functionality for a user to import and remove modules with a wide variety of options. The opinionated pattern of Switch-AzureMode has been confusing and frustrating to users. This pattern locks users into this opinionated approach to module loading. See the following quote from GitHub:

> The two "modes" in Azure are painful and confusing to work with. For example, there's two different "Get-AzureVM" cmdlets (one in each "mode"), each listing its own sets of vms (the "ServiceManagement" cmdlet lists "v1" vms and the "ResourceManager" cmdlet lists "v2" vms). This is a horrible design. Please build a PowerShell module where every module is available without having to perform confusing operations like "Switch-AzureMode": The Azure module is the only module on the planet with two modes, and for good reasons: It is a plain stupid design.

via [@trondhindenes in #428](https://github.com/Azure/azure-powershell/issues/428#issue-81097787)

There has been significant user feedback regarding the design and we have decided to make a change for the better and deprecate Switch-AzureMode.

## Future Without Switch-AzureMode

With the deprecation of Switch-AzureMode, the following changes are proposed.
- **Remove Switch-AzureMode**
 - Separate Azure and AzureResourceManager modules
 - Azure PowerShell will no longer be modal
- **Module Decomposition**
 - AzureResourceManager module will be broken up into component modules by service and function (management vs. data plane)
 - Management Examples:
    - AzureRMCompute (contains compute mgmt. cmdlets)
    - AzureRMStorage (contains storage mgmt. cmdlets)
 - Data Plane Example:
    - AzureStorage (contains cmdlets for accessing storage)
- **Cmdlet Rename**
 - All cmdlets under Azure Resource Management modules will be renamed to fit the following format: [Verb]-AzureRM[Noun]
 - Example:
    - New-AzureVm becomes New-AzureRMVm
- **Distribution Improvements**
 - PowerShell Gallery and WebPI will be used to distribute Azure PowerShell and all of its component modules.
 - This will enable quicker, independent releases much like the way Azure SDK for .Net is able to release individual NuGet packages rather than one large installer.
 - We will also be tagging modules with API versions they support.
- **Documentation Improvements and Enforcement**
 - Add build verifications to ensure documentation coverage does not decrease.
 - Automatically produce documentation upon release builds of Azure PowerShell and push the documentation to MSDN.

These changes will affect a large number of users and have the potential to break many scripts that currently use Azure Resource Management cmdlets. We chose this based on feedback from the community. In the original proposal, we decided to change the ASM cmdlet names rather than the ARM cmdlets. This met with some strong criticism. See the following quotes:

> I can't possibly begin to explain how frustrating this is.

> It seems that the decision is being made to throw bad engineering after bad engineering; introducing the Switch-AzureMode to allow the re-use of commands with different parameters and methods of use because "we liked the name".

> At least for me, September is an insane timeframe to re-work scripts that took 6 months of professional services time to develop even if the "work" is to go back and add module version checks.

> It seems to me that the right approach is to learn from the Switch-AzureMode debacle rather than build on it. Admit that the "nice names" isn't a sufficient reason to disrupt your customers and use new names like verb-AzureRMnoun for the new activities.

via [@MattAllison in #428](https://github.com/Azure/azure-powershell/issues/428#issuecomment-112912219)

> I agree. Moving the arm cmdlets to the "AzureRM"-prefix will cause the least amount of disruption of current scripts. I would move quickly tho, as some customers have no doubt already started using arm cmdlets.

via [@trondhindenes in #428](https://github.com/Azure/azure-powershell/issues/428#issuecomment-120829048)

After taking into account this [discussion](https://github.com/Azure/azure-powershell/issues/428) and the feedback from the community, we have landed on this plan of action. I think this plan will provide a solid foundation for the future of Azure PowerShell and I think the community agrees.

> @devigned Thanks. As I said above, I think this is the best of some uncomfotable choices. I look forward to it comign about swiftly and smoothly. While saying this, there is a major communication exercise to undertake - This is going to affect at least some existing scripts - and the IT Pros behind those need forewarning!

via [@doctordns in #428](https://github.com/Azure/azure-powershell/issues/428#issuecomment-126927682)

## Future Without Switch-AzureMode – Azure Automation service

The Switch-AzureMode method of Azure (ASM) and Azure Resource Manager (ARM) module use did not work in Azure’s Automation service, leading to a lack of ARM adoption in the service. The Azure PowerShell changes proposed above, along with proposed changes to Azure Automation, will make use of ASM and ARM cmdlets side by side in Azure Automation much simpler. The following change to Azure Automation is proposed:

- Azure Automation will ship out of the box the new ARM-management and data-plane modules discussed above (AzureRMCompute, AzureRMStorage, AzureStorage, etc). Since the ARM-management modules use new cmdlet names (ex: “New-AzureRMVm”), they should not conflict with the Azure (ASM) cmdlets already shipping in the service.

## The Timeline For These Changes

With the wide-ranging impact of these changes in mind, we decided to start talking about the transition many months before we intended on making this changes. We have reached the time for action and the timeline is below.

![timeline](https://cloud.githubusercontent.com/assets/386473/9155068/9957c3de-3e60-11e5-982c-46f770c74d63.png)
 
## Community Feedback And Discussion

Azure PowerShell is an open source project and as such, if you feel strongly regarding this topic and would like to have your design ideas heard, please contribute to the discussion via [GitHub issue #428](https://github.com/Azure/azure-powershell/issues/428). The Azure Automation team will also be monitoring this GitHub issue.

**Comment by pzerger** [comment]
Have to say I am very impressed with the extended level of dialogue in gathering feedback and advance notice of changes. Hate that something has to be broken, but at least you're inconveniencing the smaller group at this point. Thanks.

**Comment by devigned** [comment]
@pzerger and the rest of the community on this thread, thank you so much for caring about Azure PowerShell. It's only with direct interaction and guidance from the community that we can we hope to build a set of tools which people enjoy using.

The focus on community feedback and discussion exhibited in this thread is going to continue. We are going to do our best to make sure that any major decision that comes up about Azure PowerShell will have it's roots in open dialog amongst the community.

Thank you again to everyone contributing!

**Comment by Nepomuceno** [comment]
Sorry for getting to this discussion a little bit to late but I think that this decision will have a HUGE cost on the evolution of the Azure platform as a whole. The adoption of v2 that is already slow is going to be even slower. If you are geting the command in few months for the first time and you see "Get-AzureVM" and "Get-AzureVMRM" which one would you use ? The answer is simple, so we are going to keep hearing that the automation on AWS is a lot easier than is on Azure. And people will not use the deployment template or worst than that will get frustrated because the resources that they create with the template cannot be used by the vm that they created using the PowerShell. This is going to create a lot of frustration in the future.

I agree that this decision is the one that is going to give the smaller headache now but the cost of that for Azure in the future will be huge.

I hope that this decision can be reverted in the future and we can favor the people that are trying to go forward, and not the ones that stick with the old resources. In my humble opinion if changing the command for {verb}-AzureSM{noun} it is going to break your old scripts and the cost of coordinating that is to high just don't update we should think about the future when taking decisions about new versions and how this will be in years from now.

**Comment by doctordns** [comment]
I am going to disagree with Nepomuceno. And I'll correct him that it's the RM cmdlets that are changing , not the SM ones. That approach will mean less production automation has to be recalled and reworked to cater for this change.

The root cause of this change is the two-API monster created by MS in the first place. A more sensible approach would have been to change the SM API to accommodate resource groups, tags, etc. It would have been better to have ONLY had ONE API. But that is not what happened. This initial design decision made things bad, but the decision to create the dual module with similarly named cmdlets that are utterly incompatible just made things worse. I hope these design mistakes will serve as a great warning flag to other teams going down this road (creating multiple APIs, etc).

IMHO leaving things as they were was worse than any possible alternative. Two utterly different cmdlets with the same name contained somehow in a single module is simply bad PowerShell. It was confusing to customers and should never have happened. Fixing that mistake is going to cause pain, no matter HOW it was done -  NOT doing it would have been worse. IMO.

Leaving the older Azure cmdlets alone and only changing the ARM cmdlet names is possibly the best of a bad situation. I still thing refactoring the RM cmdlets into different modules is unnecessary and possibly confusing. All the more so when the SM cmdlets are not being refactored. But that's relatively minor given we have module autoload.

The decision as communicated to keep the SM cmdlets and redo the RM cmdlets appears to be the least bad of the alternatives. I appreciate that tough decisions had to be made and am glad they've been made and we are moving forward.

**Comment by pcgeek86** [comment]
Once again, PowerShell already handles command name conflicts. Calling a command, with a conflicting name is done as follows: `ModuleName\CommandName`. 

1. If you don't use ASM, then don't install ASM. You don't need to specify the module prefix.
2. If you don't use ARM, don't install ARM. You don't need to specify the module prefix.
3. If you use both ARM and ASM, then use module prefixes.
4. As a best practice, use module prefixes regardless, to ensure the highest reliability of your code.

The Microsoft team has already picked a direction though, and are going with it. I don't like it, but it's at least a step above where we are at today. All they really had to do was break out the separate modules and install them both at the same time into `$env:PSModulePath`. The fix was pretty simple, and didn't require the renaming of any commands, but they chose to go through the effort of renaming them to something confusing. In the end, it's not my final decision, and I've made it abundantly clear what a typical PowerShell user would expect. `Get-AzureRMVM` looks pretty ugly, and I'm not excited about the lack of readability for these commands.

Cheers,
Trevor Sullivan
Microsoft MVP: PowerShell
http://trevorsullivan.net
http://twitter.com/pcgeek86

**Comment by doctordns** [comment]
Trevor - I agree to your 1st and 2nd points.  As for point 3 - I'm not convinced forcing module name prefixes is a great solution - it makes long command lines even longer! 
IMO, what should haver happened was a simpler prefix (ASM and ARM) could have been adopted from the get go with sepearate modules. I agree with you about the ugliness and less than outstanding readability. But I do think the way proposed is better than what we have now.

**Comment by robplank** [comment]
I agree with Trevor, only install the modules you need and if using both modules use the module prefix should have been the solution, it works even in Azure Automation. I just wish we could get the new module and move past this issue and ensure all of the cmdlets have the proper documentation.

**Comment by alexandair** [comment]
Don't forget that there are scenarios when you need to use both ASM- and ARM-based cmdlets, just because some of the functionalities are still not implemented as ARM cmdlets.

**Comment by Nepomuceno** [comment]
The question is actually quite simple is about how the things will look in
6 months. A lot of the current services are going to be deprecates like the
cloud services. Think How it is going to be in 1 year when someone that is
starting with azure and see 2 different commands New-AzureVM and
New-AzureRMVM which one they will choose?

**Comment by pcgeek86** [comment]
@Nepomuceno +1, if they're going to destroy the command names, they should destroy the ones that are going to be deprecated, rather than ruining the "new" user experience.

Cheers,
Trevor Sullivan
Microsoft MVP: PowerShell

**Comment by devigned** [comment]
Well, after the old ones are gone what stops us from aliasing the ARM cmdlets back to the "Azure" names :)?

**Comment by Nepomuceno** [comment]
If we want to go back after we are going to be in danger of breaking
compatibility.  We should think in the future does who don't want to change
their old script just don't update the azure version

**Comment by devigned** [comment]
@Nepomuceno there are many options. That certainly is one that would work.

**Comment by pcgeek86** [comment]
@devigned Nothing stopping you from aliasing them in the future, but you might want to consider the confusion to end users before making such decisions.

**Comment by devigned** [comment]
@pcgeek86 any major choices will have the community involved.

**Comment by jodoglevy** [comment]
@Nepomuceno in terms of how they'd choose between New-AzureVM and New-AzureRMVM, its the same as how they choose between New-AzureVM (ASM) and New-AzureVM (ARM) today. Except now the names will be different, so it will be less confusing in documentation and samples. On top of that, if they are just starting out with Azure, they should use New-AzureRMVM since it is the new V2 VM model in Azure. If they don't know to do that then it is a failure in our documentation, and not related to cmdlet names, since regardless of cmdlet name they are picking the old model. I think what might help is if the Azure PowerShell installer let you choose whether to install the ASM cmdlets, the ARM cmdlets, or both, and for ASM labeled them as "V1, deprecated" and for ARM as "V2, recommended". Same on PSGallery. @devigned may be something to consider.

@pcgeek86 don't you think calling the adding of "RM" to cmdlet names as "destroying" and "ruining" them is a little much? You are a PowerShell MVP, so you know about prefixing. Most people do not know or use prefixing in their PowerShell scripts, because cmdlet names are almost always unique, which they should be. Prefixing is just a get out of jail free card. Can you suggest a better prefix? Maybe "AzureV2" instead of "AzureRM"?

**Comment by pcgeek86** [comment]
I don't really think it's a "little much." I think it's an accurate assessment of how they are today, versus what they are becoming. If you tried doing the same thing in a PowerShell module that was much more popular with other PowerShell MVPs, you would have a similar uproar. The Azure PowerShell module is naturally benefitting from the lack of input from most PowerShell MVPs, who aren't using Azure.

I don't have to be a PowerShell MVP to know about module prefixing. It has been around for 10 years, and is simply a matter of training. Trying to solve a training problem with a technical solution is an exercise in disaster. I can't suggest a better noun prefix; I can only suggest using module prefixing, as defined by the PowerShell language specification.

Cheers,
Trevor Sullivan
Microsoft MVP: PowerShell

**Comment by larrywa** [comment]
To the Azure PowerShell team, where do you stand in regard to the estimated dates you had listed above, ie 'Release PowerShell 0.9.9 ~ 8/25'. Just trying to prep some customers for the change.

**Comment by devigned** [comment]
@larrywa we are targeting 10/01 for release at this point. We will still include 0.9.8 in the WebPI feed, so teams are not required to install the breaking changes. There will be a second feed item for this release.

**Comment by devigned** [comment]
Update is Oct. 8th release.

**Comment by SamirFarhat** [comment]
I'm really surprised why does Microsoft reports each time the new change. You should know that there are customers using Powershell and Resource Manager on production, and we spend a lot of time to prepare the switch. And then, 1 hour/day before the change release, we are informed of the report time. Please, we support Microsoft because they follow Enterprise processes which respect the customers needs.

**Comment by pcgeek86** [comment]
@SamirFarhat To the defense of the Microsoft Azure PowerShell team, they have taken a leading approach to inviting and accepting customer feedback. In this particular case, they have not rushed into anything, and have been very solid about communicating with their customers, partners, and MVPs. I think that, in this case, the feedback is somewhat misdirected. There are plenty of other Azure feature teams rushing out features and functional changes, that your feedback would probably be better served to.

Honestly, the Azure PowerShell team is doing some of the best work out of the entire Azure platform, in terms of customer engagement. I still think that the Azure identity team (Travis Gross' team) is the absolute leader, and Azure Automation (Joe Levy's team) closely follows, but David and team are doing great work.

Cheers,
**Trevor Sullivan**
Microsoft MVP PowerShell
http://trevorsullivan.net
http://twitter.com/pcgeek86

**Comment by SamirFarhat** [comment]
Hi Trevor,
I was not complaining about the MS teams works, or approaches. On the contrary, i've been always their defender faced to angry customers. I know that they are spending all the efforts to provide the best solutions, and that they are giving  possibility to the community to present  ideas and feedback. In the meanwhile, when we announce that the switch to the new Azure mode will be on October 1st, we prepared everything for this (my customer is using ARM scripts), but few days before, we change the decision and report back to 1 week further. That made me think that even the next time, the change may not happen, and so on. I was hoping just more precise dates, that's it.

**Comment by trondhindenes** [comment]
Valid point Samir. MS (although the do their best I'm sure) has a pretty bad track record when it comes to delivering on promised dates. We've given them a ton of feedback on this and communications in general needs improving, so we can only assume they listen.

**Comment by sakshi27** [comment]
Hi people, 

I wanted to know few things-
•	Will 9.8 version continue to support Switch-AzureMode cmdlet or not? If yes, till when if anyone knows?
•	My understanding is - Switch-AzureMode functionality will stop working only if I upgrade to 9.9 and would work properly if I continue to use 9.8 version. Please correct me if I am wrong.
•	Are there any other cmdlets getting changed with 9.9 version?

Hoping this is the right platform to ask these questions!

**Comment by patrik996** [comment]
Oct 8th is now passed, any release coming soon?

**Comment by doctordns** [comment]
Sakshi:
The 0.9.8 release was put out about a month ago. It contains the Switch-AzureMode although it gives a warning message about being a depricated cmdlet. 

While with GitHub one can update a version, the tradition in this project is that each changed version is a new module. So anything that works in that version will continue to work forever.

OF course, when I say forever - I mean the client side behaviour should not change. Remember that the cmdlets are a wrapper around the 'real' API, which is REST based. This API can and sometimes does change - and those changes may render features in 0.9.8 no longer usable or compatible with the updated API. While I would not expect that to happen - Azure is evolving at 'Internet speed' and with that haste comes issues - so never say never.

WIth that said, you should be safe with 0.9.8 for some time. I would expect that should I be wrong in that assertion, we'd have plenty of warning and explanation here well in advance.

**Comment by markcowl** [comment]
The new version is released: https://github.com/Azure/azure-powershell/releases/tag/v1.0.0-preview-gallery-October2015

**Comment by markcowl** [comment]
Closing (!!!!)
